odl/solvers/nonsmooth/adupdates.py

```diff
@@ -0,0 +1,210 @@
+        \min_x \sum_{i=1}^m g_i(L_i x),
+
+    where :math:`g_i` are proper, convex and lower semicontinuous functions
+    and :math:`L_i` are linear, continuous operators for
+    :math:`i = 1, \ldots, m`. In an outer iteration, the solution is found
+    iteratively by an iteration
+
+    .. math::
+        x_{n+1} = \mathrm{arg\,min}_x \sum_{i=1}^m g_i(L_i x)
+            + \frac{\mu}{2} \|x - x_n\|^2
+
+    with some :math:`\mu > 0` according to the proximal point algorithm.
+    In the inner iteration, dual variables are introduces for each of the
+    components of the sum. The Lagrangian of the problem is given by
+
+    .. math::
+        S_n(x; v_1, \ldots, v_m) = \sum_{i=1}^m (\langle v_i, L_i x \rangle
+            - g_i^*(v_i)) + \frac{\mu}{2} \|x - x_n\|^2.
+
+    Given the dual variables, the new primal variable :math:`\tilde x_{n+1}`
+    can be calculated by directly minimizing :math:`S_n` with respect to
+    :math:`x`. This corresponds to the formula
+
+    .. math::
+         \tilde x_{n+1} = x_n - \frac{1}{\mu} \sum_{i=1}^m L_i^* v_i.
+
+    The dual updates are executed according to the following rule:
+
+    .. math::
+        v_j^+ = \mathrm{Prox}^{\mu M_j^{-1}}_{g_j^*}
+            (v_j + \mu M_j^{-1} L_j \tilde x_{n+1}),
+
+    where :math:`M_j` is a diagonal matrix with positive diagonal entries such
+    that :math:`M_j - L_j L_j^*` is positive semidefinite. The matrix
+    :math:`M_j` is incorporated into this function via the variable `majs` in
+    one of the following ways:
+
+    * Setting `majs[i]` a positive float :math:`\gamma` corresponds to the
+    choice :math:`M_i = \gamma \mathrm{Id}`.
+
+    * Assume that `g_j` is a `SeparableSum`, then setting `majs[i]` a list
+    :math:`(\gamma_1, \ldots, \gamma_k)`of positive floats corresponds to the
+    choice of a block-diagonal matrix :math:`M_j`, where each block corresponds
+    to one of the space components and equals :math:`\gamma_i \mathrm{Id}`.
+
+    * Assume that `g_j` is a `L1_norm` or a `L2_norm_squared`, then setting
+    `majs[i]` a `g_j.domain.element` :math:`z` corresponds to the choice
+    :math:`\mathrm{diag}(z)`.
+
+    References
+    ----------
+    [MF2015] McGaffin, M G, and Fessler, J A. *Alternating dual updates
+    algorithm for X-ray CT reconstruction on the GPU*. IEEE Transactions
+    on Computational Imaging, 1.3 (2015), pp 186--199.
+    """
+    # Check the lenghts of the lists (= number of dual variables)
+    length = len(g)
+    if not len(L) == length:
+        raise ValueError('Number of `L` {} and number of `g` {} do not agree.'
+                         ''.format(len(L), length))
+
+    if not len(majs) == length:
+        raise ValueError('Number of `majs` {} and number of `g` {} do not'
+                         ' agree.'.format(len(majs), length))
+
+    # Check if operators have a common domain
+    # (the space of the primal variable):
+    domain = L[0].domain
+    if any(domain != opi.domain for opi in L):
```

Comment:
prefer swapping the order here to follow natural language, e..g. you would typically not say "if the expected domain is not equal to the domain", rather you would say "if the domain is not equal to the expected domain".

In code,

```python
if any(opi.domain != domain for opi in L):
```